airbyte-integrations/connectors/source-relational-db/src/main/java/io/airbyte/integrations/source/relationaldb/StateDecoratingIterator.java

```diff
@@ -60,36 +68,42 @@ private String getCursorCandidate(final AirbyteMessage message) {
 
   @Override
   protected AirbyteMessage computeNext() {
-    if (intermediateStateMessage != null) {
-      final AirbyteMessage message = intermediateStateMessage;
-      intermediateStateMessage = null;
-      return message;
-    } else if (messageIterator.hasNext()) {
-      recordCount++;
+    if (messageIterator.hasNext()) {
+      if (emitIntermediateState && intermediateStateMessage != null) {
+        final AirbyteMessage message = intermediateStateMessage;
+        intermediateStateMessage = null;
+        emitIntermediateState = false;
+        return message;
+      }
+
+      totalRecordCount++;
       final AirbyteMessage message = messageIterator.next();
       if (message.getRecord().getData().hasNonNull(cursorField)) {
         final String cursorCandidate = getCursorCandidate(message);
-        if (IncrementalUtils.compareCursors(maxCursor, cursorCandidate, cursorType) < 0) {
+        final int cursorComparison = IncrementalUtils.compareCursors(maxCursor, cursorCandidate, cursorType);
+        if (cursorComparison < 0) {
```

Comment:
How come we're breaking this logic into a `cursorComparison` variable that seems to be used only once here instead of keeping the old logic of 

```java
if (IncrementalUtils.compareCursors(maxCursor, cursorCandidate, cursortype) {}
```

Reply:
This branch is separated from another one I am working on. In that one, I also need to handle the case when `cursorComparison == 0`.

Reply:
Reverted.